### DISCUSSION TITLE
Simple closed-form solution for alpha in detect_clearsky

### DISCUSSION THREAD

**Discussion by agodbehere** [pvlib/pvlib-python#2171] [Ideas]
The algorithm in clearsky.detect_clearsky optimizes for a scaling parameter alpha by finding a value that minimizes the root mean-squared error (RMSE) between measurements and a scaled clear-sky model. Currently, the implementation computes the RMSE and uses scipy.optimize to iteratively minimize the function. However, a simple closed-form solution exists, bypassing the need to call scipy.optimize.

The objective function is $f(\alpha) = \sqrt{\frac{\sum_i (G_i - \alpha C_i)^2}{n}}$, where $G$ represents measurements of GHI and $C$ represents the clear-sky model.

Since $f(\alpha)$ is nonnegative, $\arg\min_\alpha f(\alpha) = \arg\min_\alpha f^2(\alpha)$. The value $n$ is a constant and similarly does not affect the solution. 

Thus, we end up with a modified objective function $\hat{f}(\alpha) = \sum_i (G_i-\alpha C_i)^2$. Solving for $\frac{d\hat{f}}{d\alpha} = 0$ yields the optimal value: $\hat{\alpha} = \frac{\sum_i G_i^2}{\sum_i G_iC_i}$. The result is nonnegative, so the constraint $\alpha > 0$ is satisfied implicitly.

**Comment by cwhanse** [comment]
Thanks for pointing this out. The implementation went through several iterations and this slipped by.

Could you check that formula for the minimum? I get $\sum GC / \sum C^2$

**Reply by agodbehere** [reply]
You are correct!

In "Identification of Periods of Clear Sky Irradiance in Time Series of GHI Measurements" (Reno & Hansen), they have $\alpha$ scaling $G$ (the measurements) instead of $C$ (the model), while the code scales $C$ instead. I mixed things up when converting to the code's perspective.

**Reply by cwhanse** [reply]
I had forgotten about that typo in the 2016 paper. The intent is to scale the clear-sky model by $\alpha$, not to scale the GHI measurements.

Are you interested to submit a pull request to replace the call to scipy.optimize?

**Comment by agodbehere** [comment]
I would be happy to do so. Thank you!
